PRFC: Scaffolder permissions

vinzscam opened the pull request:
# Need
The permission framework was introduced to provide authorization to catalog entities, actions and can also be used to toggle the visibility of certain UI elements using React components and hooks. The scaffolder automatically generates a UI with actions and form elements based on a YAML schema defined by users, where the result is to run tasks defined inside the template. 

The content of a template is defined inside its own catalog entity file (`catalog-info.yaml`). Due to the ability for the catalog to be integrated with the permission framework it’s possible to hide specific templates to certain users. However, once a template (entity) is visible, its entire content will also be visible. This could be a problem in case a Backstage instance is used by multiple tenants with varying levels of permissions. Allowing granular control over who can utilize certain form fields, sections and steps would help scaling the adoption of a template.

Here is the [original issue](https://github.com/backstage/backstage/issues/11403) detailing the need for the permission framework to work with the scaffolder templates so Backstage admins can define more granular control over authorizing access to fields, sections and steps within a scaffold template.

# Proposal

We are proposing to treat the parts of a scaffolder template config as a resource in the context of the permission framework that allows a policy to be written matching annotations on the subparts of a scaffolder template. This involves extending the template schema by adding an optional “metadata” property onto each of the subparts: “parameters”, “properties” and “steps”. The example code attached reuses the concept of `tags`. Adding a tag to one of the sections of the template permits categorizing them in order to be targeted using conditional decisions of the permission framework.
Let’s take this template as an example:

```yaml
apiVersion: scaffolder.backstage.io/v1beta3
kind: Template
metadata:
  name: my_custom_template
  parameters:
    - title: Provide some simple information
       metadata:
         tags:
           - bang
      properties:
        description:
          title: Description
          type: string
            metadata:
              tags:
                - bar
  steps:
    - id: step-one
      name: First log
      action: debug:log
      input:
        message: hello
      metadata:
        tags:
          - foo
    - id: step-two
      name: Second log
      action: debug:log
      input:
        message: world
```

The template contains two steps, with step-one annotated with the foo metadata.

```typescript
class ExamplePermissionPolicy implements PermissionPolicy {
  async handle(
    request: PolicyQuery,
    _user?: BackstageIdentityResponse,
  ): Promise<PolicyDecision> {
    if (
      isResourcePermission(
        request.permission,
        RESOURCE_TYPE_SCAFFOLDER_TEMPLATE,
      )
    ) {
      return createScaffolderConditionalDecision(
        request.permission,
        scaffolderConditions.hasTag({ tag: 'foo' }),
      );
    }

    return {
      result: AuthorizeResult.ALLOW,
    };
  }
}
```

The permission policy above will allow us to be able to execute only the steps annotated with the foo tags, meaning that step-two will never be executed by the scaffolder’s backend and neither will be visible in the frontend.

The policy above will also hide properties or parameters without the foo tags defined: In the example the property “description” will not be visible, since it doesn’t match the foo as metadata. In order to differentiate between properties, steps and parameters scaffolder exposes 3 different resource permissions:

- `templateParameterReadPermission`
- `templatePropertyReadPermission`
- `templateStepReadPermission`

All of them match `RESOURCE_TYPE_SCAFFOLDER_TEMPLATE` as resource type, sharing the same conditional rules.
So in case we want to show all the properties independently form the steps we could write a permission policy like:

```typescript
class ExamplePermissionPolicy implements PermissionPolicy {
  async handle(
    request: PolicyQuery,
    _user?: BackstageIdentityResponse,
  ): Promise<PolicyDecision> {
    if (isPermission(request.permission, templatePropertyReadPermission)) {
      return { result: AuthorizeResult.ALLOW };
    }

    if (isPermission(request.permission, templateStepReadPermission)) {
      return createScaffolderConditionalDecision(
        request.permission,
        scaffolderConditions.hasTag({ tag: foo }),
      );
    }

    return {
      result: AuthorizeResult.ALLOW,
    };
  }
```

The metadata shape we’ve mentioned above is just an example and we are open to suggestions about how this shape looks. One of the advantages of the metadata object is its extensibility, we can add to it as and when more cases arise. For example, in the case of “owner” we can add a new isOwner rule. Another option we discussed is to just use the current EntityMeta shape that contains title, description, labels, annotations, etc.

# Alternative

We explored and built an alternative approach (defined below), but ultimately decided against them as the above approach worked best with the current implementation of the permission framework.

## Capabilities

Introduce a new top-level object in the template spec for “capabilities”. This object includes a set of unique strings that represent the capabilities that will be referenced within the template. In the parameters section, capabilities can be referenced to conditionally render properties/sections. In the steps section, capabilities can be referenced to skip actions or provide default values for restricted parameters.

```yaml
apiVersion: scaffolder.backstage.io/v1beta3
kind: Template
metadata:
  name: my_custom_template
  steps:
    - id: step-one
      backstage:capabilities: ${{ capabilities.canPublish }}
      name: First log
      action: debug:log
      input:
        message: hello
    - id: step-two
      name: Second log
      action: debug:log
      input:
        message: world
```

```typescript
class ExamplePermissionPolicy implements PermissionPolicy {
  async handle(
    request: PolicyQuery,
    _user?: BackstageIdentityResponse,
  ): Promise<PolicyDecision> {
    if (isPermission(request.permission, templateSchemaExecutePermission)) {
      return createScaffolderConditionalDecision(request.permission, [
        scaffolderConditions.allowCapabilities({
          capabilities: ['canPublish'],
        }),
      ]);
    }


    return {
      result: AuthorizeResult.ALLOW,
    };
  }
}
```

Before evaluating a template to display a form or execute its steps, the scaffolder backend sends an authorization request for each of the capabilities defined in the template. These capabilities are included in the `resourceRef` field of the authorization request. The authorization results are made available during evaluation.

**Reason for not recommending**: Added complexity into rendering, no support for testing and checking conditions while authoring, would require more documentation on what is available and how these conditions should be written, expose authorization logic into the template that could be seen by using the catalog, added cognitive load on adopters by introducing a new pattern into the scaffolder and permission framework. Also, the conditional rules are not going to be composable using the operators (anyOf/allOf/not) provided by the permission framework, since the new rule isn’t returning a decision.

<!-- Please describe what you added, and add a screenshot if possible.
     That makes it easier to understand the change so we can :shipit: faster. -->

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

benjdlambert commented:
Hey buddy :wave: thanks for this!

So a few higher level thoughts here, and maybe I'm misunderstanding some things but just want to clarify.

So initially we thought that some of the existing catalog permissions would be good enough for the scaffolder. For instance you can show and hide different templates to users and therefore implicitly can run those templates, and that if you're not part of required groups for instance you can't run those templates.

What it looks from this PR however is that you want to have finer grain control over the actions that can be run given these tags, but I'm feeling that in both implementations here, we have an unfortunate problem that the permissions or tags are bound to these actions in the template context, even though the underlying action that runs is the same through different templates. What I mean by this is that there's an unfortunate problem that other templates that use the same actions but don't apply these tags, either through malicious actors registering templates or through forgetting to do it, the protection kind of dies which I'm not sure we want either.

I wonder if it makes sense to be able to define these permission at an action level itself, so when you register actions and custom actions in the backend you can define who can run these actions and with what input for instance? I'm not sure what that looks like in practice, but it feels a little more secure and a little more foolproof than having to have them defined in the manifest.

vinzscam commented:
> Hey buddy 👋 thanks for this!
> 
> I wonder if it makes sense to be able to define these permission at an action level itself, so when you register actions and custom actions in the backend you can define who can run these actions and with what input for instance? I'm not sure what that looks like in practice, but it feels a little more secure and a little more foolproof than having to have them defined in the manifest.

Thank you for the feedback!
I just want to clarify that the scope of this PRFC isn’t to enhance security on the scaffolder or to protect certain actions from being executed.
As you mentioned, the visibility of some templates can be protected by using the catalog permissions and that’s enough.
Trying to permissioned the actions is something that could make sense, since it adds yet another layer of security in the chain, but it won’t solve the issues described above, that is being able to fine grain control who should be able to see certain fields (in the frontend) and/or who should be able to execute certain **steps** (not actions).

benjdlambert commented:
Hmm, maybe for the `steps` part we could take advantage of the `if:` property then to hop over the certain steps that don't match the users group. Something like:

```yaml
steps:
  ...
  - id: something
     action: test:thing
     if: ${{ permissions.tag('dev') }}
```

And I think previously mentioned we have a `backstage:featureFlag` option I believe which turns on sections or or each step for each one. Ideally I'd like to be able to group these somehow under a common key, but I don't think that we can use `if:` there as `if/then/else` is also something part of `JSONSchema` in an upcoming release. I'm feeling that the metadata key is a little unfortunate bloat and wondering if we can simplify it somehow, or be able to restructure how we do optional things.

I wonder if we could take advantage of `if/then/else` or some of the build in syntax in `JSONSchema` in which we can compare external values not just values in the form if that makes sense rather than adding additional metadata?

Rugvip reviewed (commented):
To discuss the original idea a bit too, I'm thinking that it's very important that it's reasonably intuitive what the additional fields do.

Looking at the main bit of the example:

```yml
parameters:
  - title: Provide some simple information
    metadata:
      tags: [bang]
    properties:
      description:
        title: Description
        type: string
        metadata:
          tags: [bar]
steps:
  - id: step-one
    name: First log
    action: debug:log
    input:
      message: hello
    metadata:
      tags: [foo]
```

I don't find it immediately intuitive what the extra metadata tags do. I don't even really know that they're related to access control or any kind of security mechanism at all. I'm assuming with that that the tags aren't really meant as a generic concept that can be used for anything, but that they're there specifically to handle access control.

What do you think about something that hints a bit more that it's for access control, for example this?

```yml
parameters:
  - title: Provide some simple information
    accessControl:
      tags: [bang]
    properties:
      description:
        title: Description
        type: string
        accessControl:
          tags: [bar]
steps:
  - id: step-one
    name: First log
    action: debug:log
    input:
      message: hello
    accessControl:
      tags: [foo]
```

There's plenty of different names we could go with, and it's really just a naming exercise at that point, for example `access`, `permissions`, `restricted`, etc.

The main shift I'm encourage is to make it a bit less generic in order to make it more clear what it's for.

vinzscam commented:
> Hmm, maybe for the `steps` part we could take advantage of the `if:` property then to hop over the certain steps that don't match the users group. Something like:
> 
> ```yaml
> steps:
>   ...
>   - id: something
>      action: test:thing
>      if: ${{ permissions.tag('dev') }}
> ```
> 
> And I think previously mentioned we have a `backstage:featureFlag` option I believe which turns on sections or or each step for each one. Ideally I'd like to be able to group these somehow under a common key, but I don't think that we can use `if:` there as `if/then/else` is also something part of `JSONSchema` in an upcoming release. I'm feeling that the metadata key is a little unfortunate bloat and wondering if we can simplify it somehow, or be able to restructure how we do optional things.
> 
> I wonder if we could take advantage of `if/then/else` or some of the build in syntax in `JSONSchema` in which we can compare external values not just values in the form if that makes sense rather than adding additional metadata?

unfortunately the feature flag behave differently, since it's the user who decide to activate the steps, rather than admins.
What you are proposing here using `if: ${{ permissions.tag('dev') }}` is similar to what we have the described as "capabilities" in the Alternative section. We discarded the option because it requires to build some conditional rules which are incompatible with the permission framework. For example, in order to activate `if: ${{ permissions.tag('dev') }}`, we would need to create a rule called `activateTag()` which doesn't return a boolean. So using the rule in a policy like: 

```
return createScaffolderConditionalDecision({
  anyOf: [
    activateTag('dev'),
    activateTag('something),
    allOf([
       activateTag('else') 
    ])
  ]
})
```

won't make any sense. This is the reason why we have discarded this option in favour of the proposed one.

vinzscam commented:
> To discuss the original idea a bit too, I'm thinking that it's very important that it's reasonably intuitive what the additional fields do.
> 
> Looking at the main bit of the example:
> 
> ```yaml
> parameters:
>   - title: Provide some simple information
>     metadata:
>       tags: [bang]
>     properties:
>       description:
>         title: Description
>         type: string
>         metadata:
>           tags: [bar]
> steps:
>   - id: step-one
>     name: First log
>     action: debug:log
>     input:
>       message: hello
>     metadata:
>       tags: [foo]
> ```
> 
> I don't find it immediately intuitive what the extra metadata tags do. I don't even really know that they're related to access control or any kind of security mechanism at all. I'm assuming with that that the tags aren't really meant as a generic concept that can be used for anything, but that they're there specifically to handle access control.
> 
> What do you think about something that hints a bit more that it's for access control, for example this?
> 
> ```yaml
> parameters:
>   - title: Provide some simple information
>     accessControl:
>       tags: [bang]
>     properties:
>       description:
>         title: Description
>         type: string
>         accessControl:
>           tags: [bar]
> steps:
>   - id: step-one
>     name: First log
>     action: debug:log
>     input:
>       message: hello
>     accessControl:
>       tags: [foo]
> ```
> 
> There's plenty of different names we could go with, and it's really just a naming exercise at that point, for example `access`, `permissions`, `restricted`, etc.
> 
> The main shift I'm encourage is to make it a bit less generic in order to make it more clear what it's for.

totally agree!
We have chosen `metadata` to have something more generic in order to be reused by other features in the future.
We also would like to reuse the same concept used today for matching the entities, instead of creating a new one. Today we don't have a `accessControl` section in the entities. Instead, the permissions are applied by looking at some generic attributes stored in the entity, mainly metadata.

However, we are happy to revisit the naming of the `metadata` section 😀

vinzscam commented:
> > To discuss the original idea a bit too, I'm thinking that it's very important that it's reasonably intuitive what the additional fields do.
> > Looking at the main bit of the example:
> > ```yaml
> > parameters:
> >   - title: Provide some simple information
> >     metadata:
> >       tags: [bang]
> >     properties:
> >       description:
> >         title: Description
> >         type: string
> >         metadata:
> >           tags: [bar]
> > steps:
> >   - id: step-one
> >     name: First log
> >     action: debug:log
> >     input:
> >       message: hello
> >     metadata:
> >       tags: [foo]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I don't find it immediately intuitive what the extra metadata tags do. I don't even really know that they're related to access control or any kind of security mechanism at all. I'm assuming with that that the tags aren't really meant as a generic concept that can be used for anything, but that they're there specifically to handle access control.
> > What do you think about something that hints a bit more that it's for access control, for example this?
> > ```yaml
> > parameters:
> >   - title: Provide some simple information
> >     accessControl:
> >       tags: [bang]
> >     properties:
> >       description:
> >         title: Description
> >         type: string
> >         accessControl:
> >           tags: [bar]
> > steps:
> >   - id: step-one
> >     name: First log
> >     action: debug:log
> >     input:
> >       message: hello
> >     accessControl:
> >       tags: [foo]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > There's plenty of different names we could go with, and it's really just a naming exercise at that point, for example `access`, `permissions`, `restricted`, etc.
> > The main shift I'm encourage is to make it a bit less generic in order to make it more clear what it's for.
> 
> totally agree! We have chosen `metadata` to have something more generic in order to be reused by other features in the future. We also would like to reuse the same concept used today for matching the entities, instead of creating a new one. Today we don't have a `accessControl` section in the entities. Instead, the permissions are applied by looking at some generic attributes stored in the entity, mainly metadata.
> 
> However, we are happy to revisit the naming of the `metadata` section 😀

@Rugvip we renamed `metadata` to `backstage:accessControl` on both sections, parameters and steps.
We have also removed the possibility to exclude properties from the template (so only parameters and steps are now allowed).

benjdlambert commented:
Thinking that we're probably gonna hold off merging this until after the mainline release next week as there's a lot going into that release and want to probably try this on a next version or two before releasing. make sense? :pray:

vinzscam commented:
> Thinking that we're probably gonna hold off merging this until after the mainline release next week as there's a lot going into that release and want to probably try this on a next version or two before releasing. make sense? 🙏

yep no rush thanks! 😄

jennifersheppard-bah commented:
Hi all, apologies on the delayed comments here. 
As we discussed previously, the proposed solution would accomplish much of what I am looking for -

1. It enables my business case of hiding steps from unauthorized users - specifically steps that have business or cost ramifications if used without authorization (triggering CICD builds, deployments, etc)
2. It enables granular access control, per template and per step
3. The configuration is easy, flexible, and not coupled directly into custom actions

The only item not addressed with this solution is the security around _who_ is allowed to configure the permissions. 
By putting the configuration into the template, which will largely be controlled by a combination of a dedicated investment team and internal contributors from across an organization, I am introducing an area of risk. To provide an example, if we have a template that has our `foo` tag and an RBAC policy that says anyone in a set of groups can see and invoke `My Deployment Step`, then a developer could maliciously or accidentally update a template and remove the `foo` tag. This would remove the authorization control I put in place to protect that step. Even if it's not a malicious action, it still breaks my governance model. We can put some mitigations in place with required code reviews, branch protections, processes, and perhaps a pipeline validation step, but these are all fragile solutions that would likely break down over time. 

If there were a way to create an RBAC policy that controls access to the step based on template + step + permitted groups, that would reduce the vulnerability to just those that have permission to configure the policies. However, I can't think of a way to define this policy. If you utilize template name + step name, we're in the same situation where changing the step name in the template accidentally would impact the accessibility. This would probably be even more error-prone since the configuration isn't right alongside the step. The only option I have thought through is some kind of template inheritance where a template contract (and permissions) can be defined that implementing templates must adhere to. This would enforce the access control separately from the template itself. However, this is only a theoretical option based on my experience with API RBAC solutions. 

The solution proposed here would be valuable to me and my users regardless. I see this as something equatable to "frontend validation", and I will need to build in additional security to protect the resources regardless of the Backstage approach, especially when dealing with automated deployments. I do think this first layer of access control will be helpful and extensible.

benjdlambert commented:
@Rugvip me and @vinzscam spoke briefly about this last week and I brought up the naming of `backstage:accessControl` and how it should probably be `backstage:permissions` or something as nothing in the current OSS is called `accessControl` and could be potentially confusing. What's your thoughts?

Rugvip commented:
@benjdlambert no strong opinion tbh, I think `backstage:permissions` seems good

vinzscam commented:
> @benjdlambert no strong opinion tbh, I think `backstage:permissions` seems good

alright PR rebased and ready to go

Rugvip commented on plugins/permission-node/src/integration/createPermissionIntegrationRouter.ts:
```diff
@@ -160,6 +161,30 @@ const applyConditions = <TResourceType extends string, TResource>(
+ *
+ * @public
+ */
+export const createIsAuthorized = <TResource, TQuery>(
```
Do you think it would make sense to more clearly pair this up with `createConditionTransformer`? From my understanding, this is the immediate version where we have the resource, while the condition transformer turn the conditions into the resource-specific query. I'm thinking something like `createConditionApplicator`?

vinzscam replied:
`createAuthorizer`?
I honestly would prefer to expose `createGetRule` method and keep `createIsAuthorized` as scaffolder internal. I feel it's hard to explain what `createIsAuthorized` is supposed to do

Rugvip replied:
If we export `createGetRule` we'd also need to export `applyConditions` right? Feels like those two are a bit lower level and it makes sense to bundle them together into a single utility tbh.

This is all about conditions still though right? So I'm feeling that keeping the name of the form `createCondition*` would be desirable, especially with the existence of `createConditionTransformer`.

vinzscam replied:
I buy `createConditionAuthorizer` 💰

Rugvip reviewed (changes requested):
Nice! 👍 

Couple of thoughts, and then we'll need some changesets too 😁

benjdlambert commented on plugins/scaffolder-common/src/TemplateEntityV1beta3.ts:
```diff
@@ -73,6 +67,38 @@ export interface TemplateEntityV1beta3 extends Entity {
+ *
+ * @public
+ */
+export interface TemplateParameter extends JsonObject {
```
@vinzscam I think you missed the `s`. 

`TemplateParameterV1beta3` -> `TemplateParametersV1beta3`

vinzscam replied:
renamed 😄

benjdlambert commented on plugins/scaffolder-backend/api-report.md:
```diff
@@ -642,8 +648,12 @@ export interface RouterOptions {
+  // (undocumented)
   reader: UrlReader;
   // (undocumented)
+  rules?: TemplatePermissionRuleInput[];
```
I think that this is a bit generic for the `scaffolder-backend` `Router`. Let's go with `permissionRules` like the `CatalogBuilder`

https://github.com/backstage/backstage/blob/91cb7915994983167a5849bafe0badb9939ccf17/plugins/catalog-backend/src/service/CatalogBuilder.ts#L392-L399

vinzscam replied:
agree

benjdlambert reviewed (commented):
Looking pretty good - just some small bits and I think we're good to go!

vinzscam commented on plugins/scaffolder-common/alpha-api-report.md:
```diff
@@ -0,0 +1,21 @@
+export const scaffolderPermissions: ResourcePermission<'scaffolder-template'>[];
+
+// @alpha
+export const templateParameterReadPermission: ResourcePermission<'scaffolder-template'>;
```
yeah their internal `name` is different. It is used to target parameters rather than steps in the permission policy

vinzscam commented on plugins/scaffolder-common/src/TemplateEntityV1beta3.ts:
```diff
@@ -73,6 +67,38 @@ export interface TemplateEntityV1beta3 extends Entity {
+ *
+ * @public
+ */
+export interface TemplateEntityStepV1beta3 extends JsonObject {
```
JsonObject needs to stay here otherwise it will break `TemplateEntityV1beta3 extends Entity` which requires `spec` to be `JsonObject`

benjdlambert replied:
I really don't understand why though? As it worked previously and the previous steps didn't extend JSONObject?
